### PR TITLE
[Doc] Split radix_sort doc page into key and key-value API

### PR DESCRIPTION
**Pull request by dmitriy-sobolev**
Separate `radix_sort` and `radix_sort_by_key`.

Pros:

- Improved readability: no branching for a specific case, less number of interfaces per page.
- Examples do not take too much space.

Cons:
- Duplication, more to maintain. 

I assume that benefits outweigh drawbacks specifically from a user perspective.

This PR was motivated by this discussion: https://github.com/oneapi-src/oneDPL/pull/1583#discussion_r1601293197

### CODE REVIEW

File: documentation/library_guide/kernel_templates/esimd/radix_sort_by_key.rst [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+
+   // Sort a sequence of keys and values out-of-place
+   template <bool IsAscending = true, std::uint8_t RadixBits = 8,
+             typename KernelParam, typename KeysIterator1,
+             typename ValsIterator1, typename KeysIterator2,
+             typename ValsIterator2>
+   sycl::event
+   radix_sort_by_key (sycl::queue q, KeysIterator1 keys_first,
+                      KeysIterator1 keys_last, ValsIterator1 vals_first,
+                      KeysIterator2 keys_out_first, ValsIterator2 vals_out_first,
```
Comment (akukanov) [inline documentation/library_guide/kernel_templates/esimd/radix_sort_by_key.rst:43]: Please change `vals` to `values` across the whole page.

Comment (dmitriy-sobolev) [reply documentation/library_guide/kernel_templates/esimd/radix_sort_by_key.rst]: Done

File: documentation/library_guide/kernel_templates/esimd/radix_sort.rst [hunk 2]
```diff
@@ -1,37 +1,37 @@
...
----------------------------------------------------
+-----------------------------
+radix_sort Function Templates
+-----------------------------
 
-The ``radix_sort`` and ``radix_sort_by_key`` functions sort data using the radix sort algorithm.
+The ``radix_sort`` function sorts data using the radix sort algorithm.
 The sorting is stable, ensuring the preservation of the relative order of elements with equal keys.
 The functions implement a Onesweep* [#fnote1]_ algorithm variant. Both in-place and out-of-place
 overloads are provided. For out-of-place overloads, the input data order is preserved.
```
Comment (akukanov) [inline documentation/library_guide/kernel_templates/esimd/radix_sort.rst:11]: I do not understand why "data order" is mentioned; it's a bit confusing. I believe the intended message is that the input data sequence is not affected by sorting.
For example, the documentation of CUB states it much clearer:
> The contents of the input data are not altered by the sorting operation.

File: documentation/library_guide/kernel_templates/esimd/radix_sort.rst [hunk 3]
```diff
@@ -1,37 +1,37 @@
...
    namespace oneapi::dpl::experimental::kt::gpu::esimd {
 
-   // Sort a single sequence
+   // Sort in-place
    template <bool IsAscending = true, std::uint8_t RadixBits = 8,
              typename KernelParam, typename Iterator>
    sycl::event
    radix_sort (sycl::queue q, Iterator first, Iterator last,
-               KernelParam param); // (1)
+               KernelParam param) // (1)
```
Comment (akukanov) [inline documentation/library_guide/kernel_templates/esimd/radix_sort.rst:26]: I think it is better to add missing semicolons than remove proper ones. Essentially, the synopsis consists of forward declarations, and the closing semicolon is required for those.

Comment (dmitriy-sobolev) [reply documentation/library_guide/kernel_templates/esimd/radix_sort.rst]: Added

File: documentation/library_guide/kernel_templates/esimd/radix_sort.rst (hunk 2 above)
Comment (dmitriy-sobolev) [reply documentation/library_guide/kernel_templates/esimd/radix_sort.rst]: I clarified it. I also simplified the summaries of both algorithms. I hope it became more clear overall.

File: documentation/library_guide/kernel_templates/esimd/radix_sort.rst [hunk 4]
```diff
@@ -1,24 +1,25 @@
...
+radix_sort Function Templates
+-----------------------------
 
-The ``radix_sort`` and ``radix_sort_by_key`` functions sort data using the radix sort algorithm.
-The sorting is stable, ensuring the preservation of the relative order of elements with equal keys.
-The functions implement a Onesweep* [#fnote1]_ algorithm variant. Both in-place and out-of-place
-overloads are provided. For out-of-place overloads, the input data order is preserved.
+The ``radix_sort`` function sorts data using the radix sort algorithm.
+The sorting is stable, preserving the relative order of elements with equal keys.
+There are in-place and out-of-place overloads. Out-of-place overloads do not alter the input sequence.
```
Comment (danhoeflinger) [inline documentation/library_guide/kernel_templates/esimd/radix_sort.rst:10]: I think I prefer the old 
"Both in-place and out-of-place overloads are provided."
to 
"There are in-place and out-of-place overloads."

However, its just my opinion, and is optional.

**Review by danhoeflinger — APPROVED** [review]
LGTM, 
I have one optional comment about wording preference, which I will leave up to you. Thanks for making this change, I think it makes it more readable, even if it requires some duplication.

**Review by danhoeflinger — APPROVED** [review]

File: documentation/library_guide/kernel_templates/esimd/radix_sort.rst [hunk 5]
```diff
@@ -301,92 +210,16 @@ radix_sort Out-of-Place Example
...
-
-   3 2 1 5 3 3
-   r o s d t e
-
-   1 2 3 3 3 5
-   s o r t e d
-
-
-.. _memory-requirements:
+.. _keys-memory-requirements:
```
Comment (akukanov) [inline documentation/library_guide/kernel_templates/esimd/radix_sort.rst:214]: Should the labels be globally unique? If yes, I suggest to use the name of the page as the prefix, i.e. `_radix-sort-memory-requirements` etc.

Comment (dmitriy-sobolev) [reply documentation/library_guide/kernel_templates/esimd/radix_sort.rst]: I've also applied to radix_sort_by_key.rst page.

**Comment by dmitriy-sobolev** [comment]
@dcbenito, could you assist with verifying the changes?

**Review by danhoeflinger — APPROVED** [review]

File: documentation/library_guide/kernel_templates/esimd/radix_sort_by_key.rst [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+|                                               |                                                                     |
++-----------------------------------------------+---------------------------------------------------------------------+
+
+
+**Type Requirements**:
+
+- The element type of sequence(s) to sort must be a C++ integral or floating-point type
+  other than ``bool`` with a width of up to 64 bits.
+
+.. note::
```
Comment (dcbenito) [inline documentation/library_guide/kernel_templates/esimd/radix_sort_by_key.rst:105]: This formatting is odd and not rendering in Note form. Should this be a note or a header?

File: documentation/library_guide/kernel_templates/esimd/radix_sort_by_key.rst [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+-------------------------
+
+Local memory is used for reordering key-value pairs within a work-group,
+and for storing internal data such as radix value counters.
+The used amount depends on many parameters; below is an upper bound approximation:
+
+   N\ :sub:`keys_per_workgroup` + N\ :sub:`values_per_workgroup` + C
+
+where N\ :sub:`keys_per_workgroup` and N\ :sub:`values_per_workgroup` are the amounts of memory
+to store keys and values, respectively.  `C` is some additional space for storing internal data.
```
Comment (dcbenito) [inline documentation/library_guide/kernel_templates/esimd/radix_sort_by_key.rst:307]: Please remove the double space between "...respectively.  'C' is some..."

**Review by dcbenito — APPROVED** [review]
Minor changes, otherwise LGTM.

File: documentation/library_guide/kernel_templates/esimd/radix_sort_by_key.rst (hunk 7 above)
Comment (dmitriy-sobolev) [reply documentation/library_guide/kernel_templates/esimd/radix_sort_by_key.rst]: Thanks, fixed.

File: documentation/library_guide/kernel_templates/esimd/radix_sort_by_key.rst (hunk 6 above)
Comment (dcbenito) [reply documentation/library_guide/kernel_templates/esimd/radix_sort_by_key.rst:105]: I think that the Current Limitations sub-header was confusing me. I built the docs and can see that it does render correctly, so this is a non-issue that you can ignore.

**Review by danhoeflinger — APPROVED** [review]

**Pull request merged**